### ISSUE TITLE
Should be using virtual methods?

### ISSUE THREAD

**Issue by matthijskooijman** [pololu/qtr-sensors-arduino#1]
While looking over the code, I found this peculiar piece of code:

```
if (_type == QTR_RC)
{
    ((QTRSensorsRC*)this)->readPrivate(sensor_values);
    emittersOff();
    if(readMode == QTR_EMITTERS_ON_AND_OFF)
        ((QTRSensorsRC*)this)->readPrivate(off_values);
}
else
{
    ((QTRSensorsAnalog*)this)->readPrivate(sensor_values);
    emittersOff();
    if(readMode == QTR_EMITTERS_ON_AND_OFF)
        ((QTRSensorsAnalog*)this)->readPrivate(off_values);
}
```

It seems that the `QTRSensors` object keeps a `_type` variable to tell the code
which subclass it is, and then based on that calls either the
`QTRSensorsAnalog` or `QTRSensorsRC` version of the `readPrivate` method.

This kind of thing is what C++ has virtual methods for. By declaring
`readPrivate` as a (pure) virtual method in `QTRSensors` and then implementing
it in the subclasses, you can just call `this->readPrivate()` and the compiler
will (generate code to) find out which subclass is being used at runtime.

Or is there some reason that using virtual methods is not an option here?

Gr.

Matthijs

**Comment by DavidEGrayson** [comment]
Hello, Matthijs.  It has been a long time, but I just wanted to say that you are absolutely right.  It would have been better to use virtual methods, and we will consider that the next time we make any substantial changes to the code.

**Comment by DavidEGrayson** [comment]
Actually, Kevin fixed this back in 2014 with commit 332742a0a231a309b513d890e25c47febddbfe6d, so I will close this issue.
